src/ray/object_manager/connection_pool.h

```diff
@@ -0,0 +1,148 @@
+#ifndef RAY_CONNECTION_POOL_H
+#define RAY_CONNECTION_POOL_H
+
+#include <algorithm>
+#include <cstdint>
+#include <deque>
+#include <map>
+#include <memory>
+#include <thread>
+
+#include <boost/asio.hpp>
+#include <boost/asio/error.hpp>
+#include <boost/bind.hpp>
+
+#include "ray/id.h"
+#include "ray/status.h"
+
+#include "format/object_manager_generated.h"
+#include "object_directory.h"
+#include "object_manager_client_connection.h"
+
+namespace asio = boost::asio;
+
+namespace ray {
+
+class ConnectionPool {
+ public:
+  /// Callbacks for GetSender.
+  using SuccessCallback = std::function<void(SenderConnection::pointer)>;
+  using FailureCallback = std::function<void()>;
+
+  /// Connection type to distinguish between message and transfer connections.
+  enum ConnectionType { MESSAGE = 0, TRANSFER };
+
+  /// Connection pool for all connections needed by the ObjectManager.
+  ///
+  /// \param object_directory The object directory, used for obtaining
+  /// remote object manager connection information.
+  /// \param connection_service The io_service the connection pool should use
+  /// to create new connections.
+  /// \param client_id The ClientID of this node.
+  ConnectionPool(ObjectDirectoryInterface *object_directory,
+                 asio::io_service *connection_service, const ClientID &client_id);
```

Comment:
Use references instead of pointers.

If we can, can `ObjectDirectoryInterface` be `const`?

Reply:
This will become a shared pointer.

Reply:
See comment in #1791 regarding `ObjectDirectory`.